pkg/operator/agent/agent.go

```diff
@@ -217,40 +221,68 @@ func (a *Agent) discoverFlexvolumeDir() string {
 		return getDefaultFlexvolumeDir()
 	}
 
-	// unmarshall to a NodeConfigKubelet
+	// unmarshal to a NodeConfigKubelet
 	configKubelet := NodeConfigKubelet{}
 	if err := json.Unmarshal(nodeConfig, &configKubelet); err != nil {
 		logger.Warningf("unable to parse node config from Kubelet: %+v", err)
-		return getDefaultFlexvolumeDir()
+	} else {
+		flexvolumeDirPath = configKubelet.ComponentConfig.VolumePluginDir
 	}
-	flexvolumeDirPath = configKubelet.ComponentConfig.VolumePluginDir
-	if flexvolumeDirPath == "" {
-		// unmarshall to a NodeConfigControllerManager
-		configControllerManager := NodeConfigControllerManager{}
-		if err := json.Unmarshal(nodeConfig, &configControllerManager); err != nil {
-			logger.Warningf("unable to parse node config from controller manager: %+v", err)
-			return getDefaultFlexvolumeDir()
-		}
+
+	if flexvolumeDirPath != "" {
+		logFlexvolumeDirSource("NodeConfigKubelet")
+		return flexvolumeDirPath
+	}
+
+	// unmarshal to a NodeConfigControllerManager
+	configControllerManager := NodeConfigControllerManager{}
+	if err := json.Unmarshal(nodeConfig, &configControllerManager); err != nil {
+		logger.Warningf("unable to parse node config from controller manager: %+v", err)
+	} else {
 		flexvolumeDirPath = configControllerManager.ComponentConfig.VolumeConfiguration.FlexVolumePluginDir
-		if flexvolumeDirPath == "" {
-			flexvolumeDirPath = getDefaultFlexvolumeDir()
-		}
 	}
 
-	logger.Infof("using flexvolume dir path: %s", flexvolumeDirPath)
-	return flexvolumeDirPath
+	if flexvolumeDirPath != "" {
+		logFlexvolumeDirSource("NodeConfigControllerManager")
+		return flexvolumeDirPath
+	}
+
+	// unmarshal to a KubeletConfiguration
+	kubeletConfiguration := KubeletConfiguration{}
+	if err := json.Unmarshal(nodeConfig, &kubeletConfiguration); err != nil {
```

Comment:
this is the new code path that only works with 1.8? After the 1.8 client changes, would you expect this isn't necessary anymore?

Reply:
yes this is the new type that only works with 1.8+.  I suspect that when we have the 1.8 client to build against, we can refer to the proper type directly.  Here's the new type in the 1.8 tree: https://github.com/kubernetes/kubernetes/blob/release-1.8/pkg/kubelet/apis/kubeletconfig/types.go#L187

Reply:
ok, i'll keep an eye on it with the 1.8 changes